lib/AST/RequirementMachine/HomotopyReduction.cpp

```diff
@@ -656,6 +715,8 @@ void RewriteSystem::performHomotopyReduction(
 
     deleteRule(ruleID, replacementPath);
   }
+
+  propagateRedundantRequirementIDs();
```

Comment:
Maybe this should be done once at the end? (performHomotopyReduction() is called three times)